iml-agent/src/action_plugins/lamigo.rs

```diff
@@ -0,0 +1,176 @@
+         ",
+        fs = c.fs,
+        mdt = c.mdt,
+        cold_pool = c.cold_pool,
+        hot_pool = c.hot_pool,
+        min_age = c.min_age,
+        user = c.user,
+        iml_socket = iml_socket
+    )
+}
+
+pub async fn create_lamigo_service_unit(c: Config) -> Result<(), ImlAgentError> {
+    let path_fmt = env::get_var("LAMIGO_UNIT_PATH");
+    let iml_socket = daemon_plugins::postoffice::socket_name(&c.mailbox); // wants SOCK_DIR
+    let path = expand_path_fmt(&path_fmt, &c)?;
+    create_lamigo_service_unit_internal(path, &c, &iml_socket)
+        .err_into()
+        .await
+}
+
+fn expand_path_fmt(path_fmt: &str, c: &Config) -> strfmt::Result<String> {
+    let mut vars = HashMap::new();
+    let min_age_str = c.min_age.to_string();
+    vars.insert("fs".to_string(), &c.fs);
+    vars.insert("mdt".to_string(), &c.mdt);
+    vars.insert("user".to_string(), &c.user);
+    vars.insert("hot_pool".to_string(), &c.hot_pool);
+    vars.insert("cold_pool".to_string(), &c.cold_pool);
+    vars.insert("min_age".to_string(), &min_age_str);
+    vars.insert("lustre_device".to_string(), &c.lustre_device);
+    vars.insert("mailbox".to_string(), &c.mailbox);
+    // path_fmt is like "/etc/systemd/system/lamigo-{fs}-{mdt}.service"
+    strfmt::strfmt(&path_fmt, &vars)
+}
+
+async fn create_lamigo_service_unit_internal<P: AsRef<Path>>(
+    file_path: P,
+    c: &Config,
+    iml_socket: &str,
+) -> std::io::Result<()> {
+    if let Some(dir_path) = file_path.as_ref().parent() {
+        fs::create_dir_all(&dir_path).await?;
+    };
+    let mut file = fs::File::create(file_path).await?;
+    let cnt = format_lamigo_unit_contents(c, iml_socket);
+    file.write_all(cnt.as_bytes()).await?;
+    Ok(())
+    // fs::write(file, cnt.as_bytes()).await
+}
+
+#[cfg(test)]
+mod lamigo_tests {
+    use super::*;
+    use insta::assert_display_snapshot;
+    use tempfile::tempdir;
+
+    #[test]
+    fn test_expand_path_fmt() {
+        let config = Config {
+            fs: "LU_TEST1".into(),
+            mdt: "MDT000".into(),
+            user: "nick".into(),
+            min_age: 35353,
+            lustre_device: "192.168.0.100@tcp0:/spfs".into(),
+            hot_pool: "FAST_POOL".into(),
+            cold_pool: "SLOW_POOL".into(),
+            mailbox: "mailbox".into(),
+        };
+        let fmt1 = "/etc/systemd/system/lamigo-{fs}-{mdt}.service";
+        assert_eq!(
+            expand_path_fmt(fmt1, &config),
+            Ok("/etc/systemd/system/lamigo-LU_TEST1-MDT000.service".to_string())
+        );
+
+        let fmt2 = "/tmp/{user}/lamigo-{fs}-{mdt}.service";
+        assert_eq!(
+            expand_path_fmt(fmt2, &config),
+            Ok("/tmp/nick/lamigo-LU_TEST1-MDT000.service".to_string())
+        );
+
+        let fmt3 = "lamigo-{unknown_value}.service";
+        assert_eq!(
+            expand_path_fmt(fmt3, &config),
+            Err(strfmt::FmtError::KeyError(
+                "Invalid key: unknown_value".to_string()
```

Comment:
You are testing what is already tested in `strfmt` itself. If we don't trust `strfmt`, maybe we shouldn't use it ;) And anyway what we are interested in is covered in the below test.

Reply:
While I believe that `KeyError` is documented and [tested](https://docs.rs/crate/strfmt/0.1.6/source/src/tests/strfmt.rs), this part can be better with `.is_err()`, so that you don't depend on arbitrary strings.